File: 121.BestTimeToBuyAndSellStock/profit.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        if not prices:
            return 0

        max = min = prices[0]
        profit = 0
        for p in prices:
            max = p > max and p or max
            if p < min:
                max = min = p
            diff = max - min
            profit = diff > profit and diff or profit

        return profit


    def maxProfitOpt(self, prices):
        minVal = float('inf')
        profit = 0
        for p in prices:
            minVal = min(minVal, p)
            profit = max(profit, p-minVal)

        return profit

if __name__ == '__main__':
    print Solution().maxProfitOpt([7,1,5,3,6,4])

File: 121.BestTimeToBuyAndSellStock/profit.cc
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        if (prices.size() == 0) {
            return 0;
        }

        int profit = 0;
        int max = prices[0], min = prices[0];
        for (int p : prices) {
            max = p > max ? p : max;
            if (p < min) {
                max = (min = p);
            }
            int diff = max - min;
            profit = diff > profit ? diff : profit;
        }

        return profit;
    }


    int maxProfitOpt(vector<int>& prices) {
        int minVal = INT_MAX;
        int profit = 0;
        for (int p : prices) {
            minVal = min(p, minVal);
            profit = max(p - minVal, profit);
        }
        return profit;
    }
};
